File: remote/src/main/java/com/kelsos/mbrc/interactors/LibraryTrackInteractorImpl.java
package com.kelsos.mbrc.interactors;

import com.google.inject.Inject;
import com.kelsos.mbrc.RemoteDatabase;
import com.kelsos.mbrc.domain.Track;
import com.kelsos.mbrc.mappers.TrackMapper;
import com.kelsos.mbrc.repository.LibraryRepository;
import com.raizlabs.android.dbflow.runtime.TransactionManager;
import java.util.ArrayList;
import java.util.List;
import rx.Observable;
import rx.schedulers.Schedulers;

public class LibraryTrackInteractorImpl implements LibraryTrackInteractor {
  public static final int PAGE_SIZE = 100;
  @Inject private LibraryRepository repository;
  @Override public Observable<List<Track>> execute(int page, int items) {
    return repository.getTracks(page * PAGE_SIZE, PAGE_SIZE)
        .subscribeOn(Schedulers.io())
        .flatMap(data -> {
          final List<Track> map = new ArrayList<>();
          TransactionManager.transact(RemoteDatabase.NAME, () -> {
             map.addAll(TrackMapper.map(data));
          });
          return Observable.just(map);
        });
  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/interactors/LibraryTrackInteractorImpl.kt
package com.kelsos.mbrc.interactors

import com.google.inject.Inject
import com.kelsos.mbrc.constants.Constants
import com.kelsos.mbrc.domain.Track
import com.kelsos.mbrc.extensions.io
import com.kelsos.mbrc.mappers.TrackMapper
import com.kelsos.mbrc.repository.library.TrackRepository
import rx.Observable
import rx.lang.kotlin.toSingletonObservable

class LibraryTrackInteractorImpl : LibraryTrackInteractor {
  @Inject private lateinit var repository: TrackRepository

  override fun execute(page: Int, items: Int): Observable<List<Track>> {
    return repository.getTracks(page * Constants.PAGE_SIZE, Constants.PAGE_SIZE)
        .io()
        .flatMap { TrackMapper.map(it).toSingletonObservable() }
  }
}
